datafusion/src/optimizer/constant_folding.rs

```diff
@@ -127,164 +124,365 @@ impl<'a> ConstantRewriter<'a> {
+            .collect::<Vec<bool>>();
+        if can_rewrite.iter().all(|f| *f) {
+            return true;
+        } else {
+            for (rewrite_expr, expr) in can_rewrite.iter().zip(args) {
+                if *rewrite_expr {
+                    if let Ok(s) = evaluate_const_expr_unchecked(expr) {
+                        *expr = Expr::Literal(s);
+                    }
+                }
+            }
+        }
+        false
+    }
+    ///This attempts to simplify expressions of the form col(Boolean) = Boolean and col(Boolean) != Boolean
+    /// e.g. col(Boolean) = Some(true) -> col(Boolean). It also handles == and != between two boolean literals as
+    /// the binary operator physical expression currently doesn't handle them.
+
+    fn binary_column_const_fold(
+        &mut self,
+        left: &mut Box<Expr>,
+        op: &Operator,
+        right: &mut Box<Expr>,
+    ) -> Option<Expr> {
+        let expr = match (left.as_ref(), op, right.as_ref()) {
+            (
+                Expr::Literal(ScalarValue::Boolean(l)),
+                Operator::Eq,
+                Expr::Literal(ScalarValue::Boolean(r)),
+            ) => {
+                let literal_bool = Expr::Literal(ScalarValue::Boolean(match (l, r) {
+                    (Some(l), Some(r)) => Some(*l == *r),
+                    _ => None,
+                }));
+                Some(literal_bool)
+            }
+            (
+                Expr::Literal(ScalarValue::Boolean(l)),
+                Operator::NotEq,
+                Expr::Literal(ScalarValue::Boolean(r)),
+            ) => {
+                let literal_bool = match (l, r) {
+                    (Some(l), Some(r)) => {
+                        Expr::Literal(ScalarValue::Boolean(Some(l != r)))
+                    }
+                    _ => Expr::Literal(ScalarValue::Boolean(None)),
+                };
+                Some(literal_bool)
+            }
+            (Expr::Literal(ScalarValue::Boolean(b)), Operator::Eq, col)
+            | (col, Operator::Eq, Expr::Literal(ScalarValue::Boolean(b)))
+                if self.is_boolean_type(col) =>
+            {
+                Some(match b {
+                    Some(true) => col.clone(),
+                    Some(false) => Expr::Not(Box::new(col.clone())),
+                    None => Expr::Literal(ScalarValue::Boolean(None)),
+                })
+            }
+            (Expr::Literal(ScalarValue::Boolean(b)), Operator::NotEq, col)
+            | (col, Operator::NotEq, Expr::Literal(ScalarValue::Boolean(b)))
+                if self.is_boolean_type(col) =>
+            {
+                Some(match b {
+                    Some(true) => Expr::Not(Box::new(col.clone())),
+                    Some(false) => col.clone(),
+                    None => Expr::Literal(ScalarValue::Boolean(None)),
+                })
+            }
+            _ => None,
+        };
+        expr
+    }
+
+    fn rewrite_const_expr(&mut self, expr: &mut Expr) -> bool {
```

Comment:
Could we maybe comment on the meaning of the returned boolean value? If I understand correctly, it indicates whether the expression can be evaluated as a constant expression, right?